Issue 177 use multi media devices

adamlorek-mq opened the pull request.

jackshen commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+import { useState, useEffect, useReducer } from 'react';
+
+export type MediaDevices = {
+  cameraList: InputDeviceInfo[];
+  microphoneList: InputDeviceInfo[];
+  streams: StreamsMap;
+  initDefaultStream: () => void;
+  addStream: ({
+    type,
+    microphone,
+    camera,
+    audioConstraints,
+    videoConstraints,
+  }: {
+    type: StreamTypes;
+    microphone?: InputDeviceInfo;
+    camera?: InputDeviceInfo;
+    audioConstraints?: MediaTrackConstraints;
+    videoConstraints?: MediaTrackConstraints;
+  }) => Promise<void>;
+  toggleAudio: (id: StreamId) => void;
+  toggleVideo: (id: StreamId) => void;
+  removeStream: (id: StreamId) => void;
+  applyConstraints: ({
+    id,
+    audioConstraints,
+    videoConstraints,
+  }: {
+    id: StreamId;
+    audioConstraints?: MediaTrackConstraints;
+    videoConstraints?: MediaTrackConstraints;
+  }) => Promise<void>;
+  reset: () => void;
+};
+
+export type Resolution = {
+  name?: string;
+  width: number;
+  height: number;
+};
```
Not PR-specific, but can I suggest implementing a new pattern where we put types in a separate file? In the interest of minimising per-file LOC, it would be better to define TS types in a colocated `./types.ts` (under the same directory) and import as necessary.

adamlorek-mq replied:
We definitely need to use this approach in the future

jackshen commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+  height: number;
+};
+
+const allResolutions: Resolution[] = [
+  {
+    name: '4K',
+    width: 3840,
+    height: 2160,
+  },
+  {
+    name: 'QHD',
+    width: 2560,
+    height: 1440,
+  },
+  {
+    name: 'FHD',
+    width: 1920,
+    height: 1080,
+  },
+  {
+    name: 'HD',
+    width: 1280,
+    height: 720,
+  },
+  {
+    name: '480p',
+    width: 854,
+    height: 480,
+  },
+  {
+    name: 'SD',
+    width: 640,
+    height: 480,
+  },
+  {
+    name: '360p',
+    width: 640,
+    height: 360,
+  },
+];
```
Similar suggestion to https://github.com/dolbyio-samples/rts-app-react-publisher-viewer/pull/190/files#r1040384622, but with constants in a colocated `./constants.ts` file.

adamlorek-mq replied:
Same as above

vincentsong commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+  microphoneList: InputDeviceInfo[];
+  streams: StreamsMap;
+  initDefaultStream: () => void;
+  addStream: ({
```
we could move the properties out from the object as individual parameters or declare the type somewhere else.
```suggestion
  addStream: ( type: StreamTypes, microphone?: InputDeviceInfo ....
```
Or 
```
type StreamOptions = {
  type: StreamThypes.....
```

adamlorek-mq replied:
I've created a type with options

vincentsong commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+            },
+          });
+        }
+      } else {
```
```suggestion
      } else if (type === StreamTypes.MEDIA)
```
Or use `switch` and `case`

vincentsong commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+        });
+      }
+    } catch (error: unknown) {
+      console.log(error);
```
we can remove this log

adamlorek-mq commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+  type: StreamTypes;
+  display: MediaStream;
+  capabilities?: {
+    camera: MediaTrackCapabilities;
```
yes

adamlorek-mq replied:
I'll leave the useCameraCapabilities hook in the codebase. It should be removed when new media device hook will be implemented.

adamlorek-mq commented on libs/use-multi-media-devices/src/index.ts:
```diff
@@ -0,0 +1,474 @@
+  state: {
+    muteAudio: boolean;
+    displayVideo: boolean;
+    codec: string | null;
```
Removed

vishalkharge commented:
@adamlorek-mq @vincentsong 
Following issues are observed:
- Microphone button does not work in the display stream for screen and window
- Same display stream can be added multiple times
- For the video stream, if the camera/microphone is turned Off and resolution is changed, then camera/microphone is automatically turned On

adamlorek-mq commented:
@vishalkharge @vincentsong 

- fixed - removed microphone button
- browser lets users start multiple streams with the same source, we can't stop this behavior
- fixed

Please check this once again

vishalkharge commented:
> @vishalkharge @vincentsong
> 
> * fixed - removed microphone button
> * browser lets users start multiple streams with the same source, we can't stop this behavior
> * fixed
> 
> Please check this once again

@vincentsong @adamlorek-mq 
Verified the fixed issues, look good to me. Let's plan to merge this PR.
